sklearn/metric_learning/nca.py

```diff
@@ -0,0 +1,216 @@
+from ..base import BaseEstimator, TransformerMixin
+
+import numpy as np
+import scipy as sp
+import scipy.misc
+import scipy.optimize
+import time
+
+import sys
+
+def nca_vectorized_oracle(X, y, params):
+    dX = X[:, None, :] - X[None, :, :]  # n_samples x n_samples x n_comp
+    outer = dX[:, :, None, :] * dX[:, :, :, None]  # n_samples x n_samples x n_comp x n_comp
+
+    n_samples, n_features = X.shape
+    n_components = params.n_components
+
+    def oracle(L):
+        L = L.reshape((n_components, n_features))
+
+        Lx = np.dot(X, L.T)  # n_samples x n_comp
+        assert Lx.shape == (n_samples, n_components)
+
+        A = Lx[None, :, :] - Lx[:, None, :]  # n_samples x n_samples x n_comp
+        assert A.shape == (n_samples, n_samples, n_components)
+
+        logp = -(A*A).sum(axis=2)  # n_samples x n_samples
```

Comment:
To be checked against `np.einsum` later on: It has the advantage of not making an extra copy and the disadvantage of often being slower.